include/data_sink.hpp

```diff
@@ -0,0 +1,930 @@
+#ifndef GNURADIO_DATA_SINK_HPP
+#define GNURADIO_DATA_SINK_HPP
+
+#include "circular_buffer.hpp"
+#include "dataset.hpp"
+#include "history_buffer.hpp"
+#include "node.hpp"
+#include "tag.hpp"
+
+#include <any>
+#include <chrono>
+
+namespace fair::graph {
+
+enum class blocking_mode { NonBlocking, Blocking };
+
+enum class trigger_observer_state {
+    Start,        ///< Start a new dataset
+    Stop,         ///< Finish dataset
+    StopAndStart, ///< Finish pending dataset, start a new one
+    Ignore        ///< Ignore tag
+};
+
+// TODO is the scope where want these?
+struct null_type {};
+
+// Until clang-format can handle concepts
+// clang-format off
+template<typename T>
+concept TriggerPredicate = requires(const T p, tag_t tag) {
+    { p(tag) } -> std::convertible_to<bool>;
+};
+
+/**
+ * For the 'Multiplexed' acquisition mode: Stateful object checking all incoming tags to control which data should be sent
+ * to the listener.
+ *
+ * A new dataset is started when the observer returns @c Start or @c StopAndStart.
+ * A dataset is closed and sent when @c Stop or @StopAndStart is returned.
+ *
+ * The observer can rely on being called with each incoming tag exactly once, in the order they arrive.
+ *
+ * Example:
+ *
+ * @code
+ * // Observer observing three possible tag values, "green", "yellow", "red".
+ * // starting a dataset when seeing "green", stopping on "red", starting a new dataset on "yellow"
+ * struct color_observer {
+ *     trigger_observer_state operator()(const tag_t &tag) {
+ *         if (tag == green || tag == yellow) {
+ *             return trigger_observer_state::StopAndStart;
+ *         }
+ *         if (tag == red) {
+ *             return trigger_observer_state::Stop;
+ *         }
+ *
+ *         return trigger_observer_state::Ignore;
+ *     }
+ * };
+ * @endcode
+ *
+ * @see trigger_observer_state
+ */
+template<typename T>
+concept TriggerObserver = requires(T o, tag_t tag) {
+    { o(tag) } -> std::convertible_to<trigger_observer_state>;
+};
+
+template<typename T>
+concept TriggerObserverFactory = requires(T f) {
+    { f() } -> TriggerObserver;
+};
+// clang-format on
+
+template<typename T>
+class data_sink;
+
+struct data_sink_query {
+    std::optional<std::string> sink_name;
+    std::optional<std::string> signal_name;
+
+    static data_sink_query
+    with_signal_name(std::string_view name) {
```

Comment:
Naming nit-pick - I'd remove `with_`, or replace with `has_` - for me, `with_` implies a non-static member function that behaves like builder pattern - returns a new instance with a certain field updated.

Reply:
Done